tortoise/tortoise.go

```diff
@@ -135,103 +135,49 @@ func (t *turtle) evict(ctx context.Context) {
-		base          *ballotInfo
+		logger = t.logger.WithContext(ctx)
+		err    error
 
-		opinion *types.Opinion
 		current = t.last.Add(1)
-		err     error
 	)
 	if conf.current != nil {
 		current = *conf.current
 	}
-
-	for lid := t.evicted.Add(1); lid.Before(current); lid = lid.Add(1) {
-		for _, ballot := range t.ballots[lid] {
-			if ballot.malicious {
-				continue
-			}
-			dis, err := t.firstDisagreement(ctx, current, ballot, disagreements)
-			if err != nil {
-				logger.With().Error("failed to compute first disagreement", ballot.id, log.Err(err))
-				continue
-			}
-			disagreements[ballot.id] = dis
-			choices = append(choices, ballot)
-		}
-	}
-
-	prioritizeBallots(choices, disagreements)
-	if len(choices) == 0 {
-		choices = append(choices, &ballotInfo{layer: types.GetEffectiveGenesis()})
-	}
-	for _, base = range choices {
-		opinion, err = t.encodeVotes(ctx, base, t.evicted.Add(1), current)
-		if err == nil {
-			break
-		}
-		logger.With().Warning("error calculating vote exceptions for ballot",
-			base.id,
-			log.Err(err),
-			log.Stringer("current layer", current),
-		)
-	}
-
-	if opinion == nil {
-		// TODO: special error encoding when exceeding exception list size
-		return nil, errNoBaseBallotFound
-	}
-	logger.With().Info("choose base ballot",
-		log.Stringer("base layer", base.layer),
-		log.Stringer("voting layer", current),
-		log.Inline(opinion),
-	)
-	metrics.LayerDistanceToBaseBallot.WithLabelValues().Observe(float64(t.last.Value - base.layer.Value))
-	return opinion, nil
-}
-
-// firstDisagreement returns first layer where local opinion is different from ballot's opinion within sliding window.
-func (t *turtle) firstDisagreement(ctx context.Context, current types.LayerID, ballot *ballotInfo, disagreements map[types.BallotID]types.LayerID) (types.LayerID, error) {
-	// using it as a mark that the votes for block are completely consistent
-	// with a local opinion. so if two blocks have consistent histories select block
-	// from a higher layer as it is more consistent.
-	consistent := ballot.layer
-	if basedis, exists := disagreements[ballot.base.id]; exists && basedis != ballot.base.layer {
-		return basedis, nil
-	}
-
-	for lvote := ballot.votes.tail; lvote != nil; lvote = lvote.prev {
-		if lvote.lid.Before(ballot.base.layer) {
-			break
-		}
-		if lvote.vote == abstain && lvote.hareTerminated {
-			t.logger.With().Debug("ballot votes abstain on a terminated layer. can't use as a base ballot",
-				ballot.id,
-				lvote.lid,
-			)
-			return types.LayerID{}, nil
+	for lid := current.Sub(1); lid.After(t.evicted); lid = lid.Sub(1) {
+		var choices []*ballotInfo
+		if lid == types.GetEffectiveGenesis() {
+			choices = []*ballotInfo{{layer: types.GetEffectiveGenesis()}}
+		} else {
+			choices = t.ballots[lid]
 		}
-		for _, block := range lvote.blocks {
-			vote, _, err := t.getFullVote(t.verified, current, block)
-			if err != nil {
-				return types.LayerID{}, err
+		for _, base := range choices {
```

Comment:
how about bad beacon? why wouldn't we want to prioritize ballots with the same beacon (to minimize data in votes)?

Reply:
data in votes is ephemeral, it can be reorganized later by receiver if it wants to minimize amount of stored data. i would rather have simple code instead of meaningless optimizations

Reply:
> data in votes is ephemeral, it can be reorganized later by receiver if it wants to minimize amount of stored data. i would rather have simple code instead of meaningless optimizations

i agree with this approach in general. but in this case, the default node implementation doesn't support reorganizing votes yet, and this optimization seems rather cost-less. but probably not important enough.